src/System.Private.CoreLib/src/System/Threading/WaitSubsystem.ThreadWaitInfo.Unix.cs

```diff
@@ -118,18 +118,40 @@ public WaitableObject[] GetWaitedObjectArray(int requiredCapacity)
                 Debug.Assert(_thread == Thread.CurrentThread);
                 Debug.Assert(_waitedCount == 0);
 
-                _waitedObjects.VerifyElementsAreDefault();
-                _waitedObjects.EnsureCapacity(requiredCapacity);
-                return _waitedObjects.Items;
+#if DEBUG
+                for (int i = 0; i < _waitedObjects.Length; ++i)
+                {
+                    Debug.Assert(_waitedObjects[i] == null);
+                }
+#endif
+
+                int currentLength = _waitedObjects.Length;
+                if (currentLength < requiredCapacity)
+                    _waitedObjects = new WaitableObject[Math.Max(requiredCapacity,
+                        Math.Min(WaitHandle.MaxWaitHandles, 4 + 2 * currentLength))];
```

Comment:
Maybe `Math.Max(2, currentLength) * 2` would work better than `4 + 2 * currentLength`. When `currentLength` is 4 and a 5-object wait is requested, the former would allocate 8 elements and the latter would allocate 12. Then when 13 elements are requested, the former would allocate 16 elements and the latter would allocate 28 elements. Doesn't matter much though. Similarly elsewhere.

Reply:
The primary purpose of this logic is to avoid pathological cases where the array grows one element at a time. I got rid of the `4 +` completely since it is not required to fix the pathological case.